c10/core/Tensor.h

```diff
@@ -18,7 +18,7 @@ class C10Tensor final {
   C10Tensor& operator=(C10Tensor&&) noexcept = default;
 
   const TensorImplPtr &impl() const & noexcept;
-  TensorImplPtr impl() && noexcept;
+  TensorImplPtr&& impl() && noexcept;
```

Comment:
Why are you doing this?

Reply:
Also, this doesn't match APIs in either caffe2::Tensor nor at::Tensor

Reply:
This seemed more consistent with `std::move` behavior. `std::move(tensor)` doesn't move the tensor yet but casts it into an rvalue, so maybe `std::move(tensor).impl()` shouldn't move the impl yet but cast it into an rvalue.

Reply:
Uhhh.... or we could tell people not to write `std::move(tensor).impl()`. Let me be straight here: you'd better have a **really** good reason for adding an an rvalue reference qualifier; this is really, really confusing for people who aren't intimately familiar with C++ move semantics. (Making `x[0] = y` work is a good reason; making `std::move(tensor).impl()` avoid moving the impl immediately is not.)